packages/core/src/extensions/Comments/CommentsPlugin.ts

```diff
@@ -0,0 +1,297 @@
+          to: Math.max(to, currentPosition.to),
+        });
+
+        if (selectedThreadId === thisThreadId) {
+          decorations.push(
+            Decoration.inline(from, to, {
+              class: "bn-thread-mark-selected",
+            })
+          );
+        }
+      }
+    });
+  });
+  return {
+    decorations: DecorationSet.create(doc, decorations),
+    threadPositions,
+  };
+}
+
+export class CommentsPlugin extends EventEmitter<any> {
+  public readonly plugin: Plugin;
+  public readonly userStore: UserStore<User>;
+
+  private pendingComment = false;
+  private selectedThreadId: string | undefined;
+
+  private emitStateUpdate() {
+    this.emit("update", {
+      selectedThreadId: this.selectedThreadId,
+      pendingComment: this.pendingComment,
+    });
+  }
+
+  /**
+   * when a thread is resolved or deleted, we need to update the marks to reflect the new state
+   */
+  private updateMarksFromThreads = (threads: Map<string, ThreadData>) => {
+    const ttEditor = this.editor._tiptapEditor;
+
+    ttEditor.state.doc.descendants((node, pos) => {
+      node.marks.forEach((mark) => {
+        if (mark.type.name === this.markType) {
+          const markType = mark.type;
+          const markThreadId = mark.attrs.threadId;
+          const thread = threads.get(markThreadId);
+          const isOrphan = !!(!thread || thread.resolved || thread.deletedAt);
+
+          if (isOrphan !== mark.attrs.orphan) {
+            const { tr } = ttEditor.state;
+            const trimmedFrom = Math.max(pos, 0);
+            const trimmedTo = Math.min(
+              pos + node.nodeSize,
+              ttEditor.state.doc.content.size - 1
+            );
+            tr.removeMark(trimmedFrom, trimmedTo, markType);
+            tr.addMark(
+              trimmedFrom,
+              trimmedTo,
+              markType.create({
+                ...mark.attrs,
+                orphan: isOrphan,
+              })
+            );
+            ttEditor.dispatch(tr);
+
+            if (isOrphan && this.selectedThreadId === markThreadId) {
+              // unselect
+              this.selectedThreadId = undefined;
+              this.emitStateUpdate();
+            }
+          }
+        }
+      });
+    });
+  };
+
+  constructor(
+    private readonly editor: BlockNoteEditor<any, any, any>,
+    public readonly threadStore: ThreadStore,
+    private readonly markType: string
+  ) {
+    super();
+
+    if (!editor.resolveUsers) {
+      throw new Error("resolveUsers is required for comments");
+    }
+    this.userStore = new UserStore<User>(editor.resolveUsers);
+
+    // Note: Plugins are currently not destroyed when the editor is destroyed.
+    // We should unsubscribe from the threadStore when the editor is destroyed.
+    this.threadStore.subscribe(this.updateMarksFromThreads);
+
+    editor.onCreate(() => {
+      this.updateMarksFromThreads(this.threadStore.getThreads());
+      editor.onSelectionChange(() => {
```

Comment:
Feels weird to register this after the create hook, why not just register in the constructor separately?

Reply:
It needs to access the TipTap state, and bc we init the comments plugin in the `BlockNoteEditor` constructor, the state is still undefined at that time. Will add a comment